File: 2411-spiral-matrix-iv/spiral-matrix-iv.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    vector<vector<int>> spiralMatrix(int m, int n, ListNode* head) {
        vector<vector<int>> ans(m,vector<int>(n,-1));
        if(!head) return ans;
        ListNode*ptr = head;
        int top=0,bottom=m-1,left=0,right=n-1;
        while(ptr){
            for(int i=left;i<=right && ptr!=NULL;i++){
                ans[top][i] = ptr->val;
                ptr = ptr->next;
            }
            top++;
            for(int i=top;i<=bottom && ptr!=NULL;i++){
                ans[i][right] = ptr->val;
                ptr = ptr->next;
            }
            right--;

            for(int i=right;i>=left && ptr!=NULL;i--){
                ans[bottom][i] = ptr->val;
                ptr = ptr->next;
            }

            bottom--;

            for(int i=bottom;i>=top && ptr!=NULL;i--){
                ans[i][left] = ptr->val;
                ptr = ptr->next;
            }
            left++;
        }
        return ans;
    }
};

File: 2411-spiral-matrix-iv/spiral-matrix-iv.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */

class Solution {
    public int[][] spiralMatrix(int m, int n, ListNode head) {
        int[][] arr=new int[m][n];
        int top=0;
        int bottom=m-1;
        int left=0;
        int right=n-1;
        ListNode curr=head;
        int r=0;
        int c=0;


        while(top<=bottom && left<=right){
          
          while(c<=right){
            if(curr!=null){
                arr[r][c]=curr.val;
            }
            else{
                arr[r][c]=-1;
            }
            if(curr!=null)curr=curr.next;
            c++;
          }
          c--;
          
          top++;
          if(top<=bottom && left<=right){
            r++;
            while(r<=bottom){
                if(curr!=null){
                    arr[r][c]=curr.val;
                }
                else{
                    arr[r][c]=-1;
                }
                 if(curr!=null)curr=curr.next;
                r++;
            }
             r--;
          }
        
         right--;
         if(top<=bottom && left<=right){
            c--;
            while(c>=left){
                if(curr!=null){
                    arr[r][c]=curr.val;
                }
                else{
                    arr[r][c]=-1;
                }
                 if(curr!=null)curr=curr.next;
                c--;
            }
            c++;
         }
           
           bottom--;
           if(top<=bottom && left<=right){
            r--;
            while(r>=top){
                if(curr!=null){
                    arr[r][c]=curr.val;
                }
                else{
                    arr[r][c]=-1;
                }
                 if(curr!=null)curr=curr.next;
                r--; 
            }
            r++;
           }
           left++;

           if(left<=right){
            c++;
           }
           
        }

        return arr;
    }
}